File: app/src/main/java/com/octavianmetta/android/myanimelistsearcher/viewModel/AnimeDetailViewModel.java
package com.octavianmetta.android.myanimelistsearcher.viewModel;

import android.arch.lifecycle.LiveData;
import android.arch.lifecycle.MutableLiveData;
import android.arch.lifecycle.ViewModel;
import android.util.Log;

import com.octavianmetta.android.myanimelistsearcher.models.anime.AnimeModel;
import com.octavianmetta.android.myanimelistsearcher.rest.APIService;
import com.octavianmetta.android.myanimelistsearcher.rest.RetrofitClient;

import io.reactivex.Observable;
import io.reactivex.Observer;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.disposables.Disposable;
import io.reactivex.schedulers.Schedulers;
import retrofit2.Retrofit;

public class AnimeDetailViewModel extends ViewModel {

    private Retrofit retrofit = RetrofitClient.getClient(APIService.BASE_URL);
    private APIService MALApi = retrofit.create(APIService.class);

    private MutableLiveData<AnimeModel> animeResults;
    private CompositeDisposable compositeDisposable = new CompositeDisposable();

    public LiveData<AnimeModel> getAnimeData(int malId){
        if(animeResults == null){
            animeResults = new MutableLiveData<>();
            loadMALAnime(malId);
        }

        return  animeResults;
    }
    public void loadMALAnime(Integer malId){
        //Dijalankan setelah search. Untuk mendapatkan data hasil search
        Observable<AnimeModel> animeModelObservable = MALApi.getAnime(malId);
        animeModelObservable.subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(new Observer<AnimeModel>() {
                    @Override
                    public void onSubscribe(Disposable d) {
                        compositeDisposable.add(d);
                    }

                    @Override
                    public void onNext(AnimeModel animeModelResponses) {
                        animeResults.postValue(animeModelResponses);
                        Log.d("Response", animeModelResponses.getTitle());

                    }

                    @Override
                    public void onError(Throwable e) {
                        e.printStackTrace();
                    }

                    @Override
                    public void onComplete() {

                    }
                });
    }
    @Override
    protected void onCleared() {
        if(compositeDisposable != null || !compositeDisposable.isDisposed())
            compositeDisposable.clear();
    }
}

File: app/src/main/java/com/octavianmetta/android/myanimelistsearcher/viewModel/AnimeDetailViewModel.kt
package com.octavianmetta.android.myanimelistsearcher.viewModel

import android.arch.lifecycle.LiveData
import android.arch.lifecycle.MutableLiveData
import android.arch.lifecycle.ViewModel
import android.util.Log

import com.octavianmetta.android.myanimelistsearcher.models.anime.AnimeModel
import com.octavianmetta.android.myanimelistsearcher.rest.APIService
import com.octavianmetta.android.myanimelistsearcher.rest.RetrofitClient

import io.reactivex.Observable
import io.reactivex.Observer
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.disposables.Disposable
import io.reactivex.schedulers.Schedulers
import retrofit2.Retrofit

class AnimeDetailViewModel : ViewModel() {

    private val retrofit = RetrofitClient.getClient(APIService.BASE_URL)
    private val MALApi = retrofit?.create(APIService::class.java)

    private var animeResults: MutableLiveData<AnimeModel>? = null
    private val compositeDisposable = CompositeDisposable()

    fun getAnimeData(malId: Int): LiveData<AnimeModel>? {
        if (animeResults == null) {
            animeResults = MutableLiveData()
            loadMALAnime(malId)
        }

        return animeResults
    }

    fun loadMALAnime(malId: Int?) {
        //Dijalankan setelah search. Untuk mendapatkan data hasil search
        val animeModelObservable = MALApi?.getAnime(malId)
        animeModelObservable!!.subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(object : Observer<AnimeModel> {
                    override fun onSubscribe(d: Disposable) {
                        compositeDisposable.add(d)
                    }

                    override fun onNext(animeModelResponses: AnimeModel) {
                        animeResults!!.postValue(animeModelResponses)
                        Log.d("Response", animeModelResponses.title)

                    }

                    override fun onError(e: Throwable) {
                        e.printStackTrace()
                    }

                    override fun onComplete() {

                    }
                })
    }

    override fun onCleared() {
        if (compositeDisposable != null || !compositeDisposable.isDisposed)
            compositeDisposable.clear()
    }
}
